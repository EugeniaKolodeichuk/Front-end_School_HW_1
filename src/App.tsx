/* eslint-disable object-curly-newline */
import { lazy, Suspense, createContext, useState } from 'react';
import { Route, Switch } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import Loader from './components/Loader/Loader';
import GlobalStyle from './styled/Global';
import { theme } from './components/Theme/StyledTheme';
import { Theme, ContextTheme } from './types/types';

const AppBar = lazy(
  () => import('./components/AppBar/AppBar' /*webpackChunkName: "app-bar"*/),
);
const Container = lazy(
  () =>
    import(
      './components/Container/Container' /*webpackChunkName: "container"*/
    ),
);
const NewsView = lazy(
  () => import('./pages/NewsView/NewsView' /*webpackChunkName: "news-view"*/),
);
const ProfileView = lazy(
  () =>
    import(
      './pages/ProfileView/ProfileView' /*webpackChunkName: "profile-view"*/
    ),
);

export const ThemeContext = createContext<ContextTheme>({
  value: 'light',
  setThemeValue: null,
});

export default function App() {
  const [themeValue, setThemeValue] = useState<keyof Theme>('light'); // "light" | "dark"
  return (
    <>
      <ThemeContext.Provider value={{ value: themeValue, setThemeValue }}>
        <ThemeProvider theme={theme[themeValue]}>
          <GlobalStyle />
          <Suspense fallback={<Loader />}>
            <Container>
              <AppBar />

              <Switch>
                <Route path="/profile/:uniqueId">
                  <ProfileView />
                </Route>

                <Route path="/">
                  <NewsView />
                </Route>
              </Switch>
            </Container>
          </Suspense>
        </ThemeProvider>
      </ThemeContext.Provider>
    </>
  );
}
