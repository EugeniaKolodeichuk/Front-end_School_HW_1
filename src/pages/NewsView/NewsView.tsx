import React, { useState, useEffect } from 'react';
import shortid from 'shortid';
import styles from './NewsView.module.css';
import Loader from '../../components/Loader/Loader';
import { getTrendingFeed } from '../../service/app';
import TrendingFeed from '../../components/TrendingFeed/TrendingFeed';
import { ITrendingFeed } from '../../types/feedTypes';

export default function NewsView() {
  const [trendingFeed, setTrendingFeed] = useState<ITrendingFeed[]>([]);
  const [currentPage, setCurrentPage] = useState(1);
  const [fetching, setFetching] = useState(true);

  const scrollHandler = (event: any): void => {
    if (
      event.target.documentElement.scrollHeight -
        (event.target.documentElement.scrollTop + window.innerHeight) <
      100
    ) {
      setFetching(true);
    }
  };

  useEffect(() => {
    document.addEventListener('scroll', scrollHandler);

    return function () {
      document.removeEventListener('scroll', scrollHandler);
    };
  }, []);

  useEffect(() => {
    try {
      getTrendingFeed(currentPage).then(data => {
        setTrendingFeed([...trendingFeed, ...data]);
        setCurrentPage(previousState => previousState + 1);
      });
    } catch (error) {
      console.error(error);
    } finally {
      setFetching(false);
    }
  }, [fetching]);

  return (
    <div className={styles.wrap} key={shortid.generate()}>
      {trendingFeed.length > 0 ? (
        trendingFeed.map(feed => (
          <TrendingFeed key={shortid.generate()} feed={feed} />
        ))
      ) : (
        <Loader />
      )}
    </div>
  );
}
