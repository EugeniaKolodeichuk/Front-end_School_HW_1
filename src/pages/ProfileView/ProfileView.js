import { useParams } from 'react-router';
import React, { useState, useEffect } from 'react';
import styles from './ProfileView.module.css';
import Loader from '../../components/Loader/Loader';
import { getUserFeed, getUserInfo } from '../../service/app';
import UserVideos from '../../components/UserVideos/UserVideos';
import UserInfo from '../../components/UserInfo/UserInfo';

export default function ProfileView() {
  const { uniqueId } = useParams();

  const [userFeed, setUserFeed] = useState([]);
  const [userInfo, setUserInfo] = useState([]);

  useEffect(() => {
    try {
      getUserInfo(uniqueId).then(data => {
        setUserInfo(data);
      });
    } catch (error) {
      console.error(error);
    }
  }, [setUserInfo]);

  useEffect(() => {
    try {
      getUserFeed(uniqueId).then(data => {
        setUserFeed(data);
      });
    } catch (error) {
      console.error(error);
    }
  }, [setUserFeed]);

  return (
    <div key={userInfo.id} data-testid="profile">
      <UserInfo
        uniqueId={userInfo.uniqueId}
        nickname={userInfo.nickname}
        avatarMedium={userInfo.avatarMedium}
        signature={userInfo.signature}
      />
      <div className={styles.wrap} key={userFeed.id}>
        {userFeed > 0 ? (
          userFeed.map(user => <UserVideos key={user.createTime} user={user} />)
        ) : (
          <Loader />
        )}
      </div>
    </div>
  );
}
