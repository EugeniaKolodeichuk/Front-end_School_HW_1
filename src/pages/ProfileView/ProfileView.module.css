.wrap {
    max-width: 1300px;
    margin: 0 auto;
    display: flex;
    justify-content: space-evenly;
    flex-wrap: wrap;
}