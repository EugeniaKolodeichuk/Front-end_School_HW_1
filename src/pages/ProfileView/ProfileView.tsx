import { useParams } from 'react-router';
import { useState, useEffect } from 'react';
import StyledWrap from './StyledProfileView';
import Loader from '../../components/Loader/Loader';
import { getUserFeed, getUserInfo } from '../../service/app';
import UserVideos from '../../components/UserVideos/UserVideos';
import UserInfo from '../../components/UserInfo/UserInfo';
import { IUserInfo } from '../../types/userTypes';

export default function ProfileView() {
  const { uniqueId } = useParams<{ uniqueId: string }>();

  const [userFeed, setUserFeed] = useState<IUserInfo[]>([]);
  const [userInfo, setUserInfo] = useState<IUserInfo | null>(null);

  useEffect(() => {
    try {
      getUserInfo(uniqueId).then(data => {
        setUserInfo(data);
      });
    } catch (error) {
      console.error(error);
    }
  }, [setUserInfo]);

  useEffect(() => {
    try {
      getUserFeed(uniqueId).then(data => {
        setUserFeed(data);
      });
    } catch (error) {
      console.error(error);
    }
  }, [setUserFeed]);
  if (!userInfo) {
    return <Loader />;
  }
  return (
    <div key={userInfo.id} data-testid="profile">
      <UserInfo
        uniqueId={userInfo.uniqueId}
        nickname={userInfo.nickname}
        avatarMedium={userInfo.avatarMedium}
        signature={userInfo.signature}
      />
      <StyledWrap key={userInfo.id}>
        {userFeed ? (
          userFeed.map((user: IUserInfo) => (
            <UserVideos key={user.createTime} user={user} />
          ))
        ) : (
          <Loader />
        )}
      </StyledWrap>
    </div>
  );
}
