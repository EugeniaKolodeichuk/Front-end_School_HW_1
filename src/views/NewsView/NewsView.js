import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import defaultImage from '../../images/default.jpg';
import styles from './NewsView.module.css';
import Loader from '../../components/Loader/Loader';
import { getTrendingFeed } from '../../service/app';

export default function NewsView() {
  const [trendingFeed, setTrendingFeed] = useState([]);
  const [currentPage, setCurrentPage] = useState(1);
  const [fetching, setFetching] = useState(true);

  const scrollHandler = event => {
    if (
      event.target.documentElement.scrollHeight -
        (event.target.documentElement.scrollTop + window.innerHeight) <
      100
    ) {
      setFetching(true);
    }
  };

  useEffect(() => {
    document.addEventListener('scroll', scrollHandler);

    return function () {
      document.removeEventListener('scroll', scrollHandler);
    };
  }, []);

  useEffect(() => {
    try {
      getTrendingFeed(currentPage).then(data => {
        setTrendingFeed([...trendingFeed, ...data]);
        setCurrentPage(previousState => previousState + 1);
      });
    } catch (error) {
      console.error(error);
    } finally {
      setFetching(false);
    }
  }, [fetching]);

  return (
    <div className={styles.wrap}>
      {trendingFeed.length > 0 ? (
        trendingFeed.map(user => (
          <div className={styles.list_noorder} key={user.id}>
            {user.video.playAddr ? (
              <video width="300px" controls="controls">
                <source src={user.video.playAddr} type="video/mp4;" />
                <track kind="captions" srcLang="en" label="english_captions" />
              </video>
            ) : (
              <Loader />
            )}
            <p>{user.author.signature}</p>
            <div className={styles.user_info}>
              {user.author.avatarThumb ? (
                <img
                  key={user.video.id}
                  width="70px"
                  src={user.author.avatarThumb}
                  alt={user.author.nickname}
                  className={styles.avatar}
                />
              ) : (
                <img
                  width="150px"
                  src={defaultImage}
                  alt={user.author.nickname}
                />
              )}
              <p>
                <Link
                  className={styles.link}
                  to={`/profile/${user.author.uniqueId}`}
                >
                  <b>{user.author.nickname}</b>
                </Link>
              </p>
            </div>
            <p>
              Comments: <b>{user.stats.commentCount}</b>
            </p>
            <p>
              Likes: <b>{user.stats.diggCount}</b>
            </p>
          </div>
        ))
      ) : (
        <Loader />
      )}
    </div>
  );
}
