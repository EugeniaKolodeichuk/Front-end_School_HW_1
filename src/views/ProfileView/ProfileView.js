import { useParams } from 'react-router';
import React, { useState, useEffect } from 'react';
import defaultImage from '../../images/default.jpg';
import styles from './ProfileView.module.css';
import Loader from '../../components/Loader/Loader';
import { getUserFeed, getUserInfo } from '../../service/app';

export default function ProfileView() {
  const { uniqueId } = useParams();

  const [userFeed, setUserFeed] = useState([]);
  const [userInfo, setUserInfo] = useState([]);

  useEffect(() => {
    try {
      getUserInfo(uniqueId).then(data => {
        setUserInfo(data);
      });
    } catch (error) {
      console.error(error);
    }
  }, [setUserInfo]);

  useEffect(() => {
    try {
      getUserFeed(uniqueId).then(data => {
        setUserFeed(data);
      });
    } catch (error) {
      console.error(error);
    }
  }, [setUserFeed]);

  return (
    <div key={userInfo.id}>
      {userInfo && (
        <div key={userInfo.uniqueId} className={styles.profile}>
          <div className={styles.list_noorder}>
            {userInfo.avatarThumb ? (
              <img
                key={userInfo.nickname}
                width="150px"
                src={userInfo.avatarMedium}
                alt={userInfo.nickname}
                className={styles.avatar}
              />
            ) : (
              <img width="150px" src={defaultImage} alt={userInfo.nickname} />
            )}
            <h2 className={styles.title}>{userInfo.nickname}</h2>
            <p>{userInfo.signature}</p>
          </div>
          <h2 className={styles.title}>
            Other posts created by {userInfo.nickname}
          </h2>
        </div>
      )}
      <div className={styles.wrap} key={userFeed.id}>
        {userFeed.length ? (
          userFeed.map(user => (
            <div key={user.createTime} className={styles.list_noorder}>
              <p>
                <b>{user.stats.playCount}</b> views
              </p>
              {user.video.originCover ? (
                <img
                  key={user.video.id}
                  src={user.video.originCover}
                  alt={user.desc}
                  className={styles.main_element}
                />
              ) : (
                <img width="150px" src={defaultImage} alt={userInfo.nickname} />
              )}

              <p>{user.desc}</p>
            </div>
          ))
        ) : (
          <Loader />
        )}
      </div>
    </div>
  );
}
