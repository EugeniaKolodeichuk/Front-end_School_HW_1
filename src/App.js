/* eslint-disable func-call-spacing */
import React, { lazy, Suspense } from 'react';
import { Route, Switch } from 'react-router-dom';
import Loader from './components/Loader/Loader';

const AppBar = lazy(() =>
  import('./components/AppBar/AppBar' /*webpackChunkName: "app-bar"*/),
);
const Container = lazy(() =>
  import('./components/Container/Container' /*webpackChunkName: "container"*/),
);
const NewsView = lazy(() =>
  import('./views/NewsView/NewsView' /*webpackChunkName: "news-view"*/),
);
const ProfileView = lazy(() =>
  import(
    './views/ProfileView/ProfileView' /*webpackChunkName: "profile-view"*/
  ),
);

export default function App() {
  return (
    <Suspense fallback={<Loader />}>
      <Container>
        <AppBar />

        <Switch>
          <Route path="/profile/:uniqueId">
            <ProfileView />
          </Route>

          <Route path="/">
            <NewsView />
          </Route>
        </Switch>
      </Container>
    </Suspense>
  );
}
