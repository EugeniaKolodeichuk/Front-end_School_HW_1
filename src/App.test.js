import React from 'react';
import { unmountComponentAtNode } from 'react-dom';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import App from './App';
import ProfileView from './pages/ProfileView/ProfileView';
import NewsView from './pages/NewsView/NewsView';

test('render App', () => {
  render(<App />);
  screen.debug();
});

/* describe('React Router', () => {
  it('should render the trends page', () => {
    const history = createMemoryHistory();
    const { container, getByTestId } = render(
      <Router history={history}>
        <RouterComponent />
      </Router>,
    );
    const navbar = getByTestId('navbar');
    const link = getByTestId('link');
    expect(container.innerHTML).toMatch('Trends');
    expect(navbar).toContainElement(link);
  });
}); */

let container = null;

beforeEach(() => {
  container = document.createElement('div');
  document.body.append(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe('ProfileView', () => {
  test('render ProfileView', () => {
    const history = createMemoryHistory();
    render(
      <Router history={history}>
        <ProfileView />
      </Router>,
      container,
    );
    expect(container).toMatchSnapshot();
  });

  test('check for a home link', () => {
    const history = createMemoryHistory();
    render(
      <Router history={history}>
        <ProfileView />
      </Router>,
    );

    expect(screen.getByText(/other posts created by/i)).toBeInTheDocument();
  });
});

describe('NewsView', () => {
  test('render NewsView', () => {
    const history = createMemoryHistory();
    render(
      <Router history={history}>
        <NewsView />
      </Router>,
      container,
    );
    expect(container).toMatchSnapshot();
  });

  test('check for a home link', () => {
    const history = createMemoryHistory();
    render(
      <Router history={history}>
        <NewsView />
      </Router>,
    );

    /* expect(screen.getByText(/Trends/i)).toBeInTheDocument(); */
  });
});
