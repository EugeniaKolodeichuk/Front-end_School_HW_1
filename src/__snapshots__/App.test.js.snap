// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsView render NewsView 1`] = `<div />`;

exports[`ProfileView render ProfileView 1`] = `<div />`;
